src/bencher/db.rs

```diff
@@ -0,0 +1,394 @@
+//! some smoothing will occur.
+
+use std::collections::VecDeque;
+use std::ops::Range;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use bytes::Bytes;
+use rand::{Rng, RngCore, SeedableRng};
+use rand_xorshift::XorShiftRng;
+use slatedb::config::WriteOptions;
+use slatedb::db::Db;
+use tokio::time::Instant;
+use tracing::info;
+
+/// How frequently to dump stats to the console.
+const STAT_DUMP_INTERVAL: Duration = Duration::from_secs(10);
+
+/// How far back to look when dumping stats.
+const STAT_DUMP_LOOKBACK: Duration = Duration::from_secs(60);
+
+/// How frequently to update stats between puts and gets and
+/// how frequently to check if we need to dump new stats.
+const REPORT_INTERVAL: Duration = Duration::from_millis(100);
+
+/// The size of each window.
+const WINDOW_SIZE: Duration = Duration::from_secs(10);
+
+/// A key generator trait that generates keys for the benchmarker.
+pub trait KeyGenerator: Send {
+    fn next_key(&mut self) -> Bytes;
+}
+
+/// A key generator that generates random keys of a fixed length.
+pub struct RandomKeyGenerator {
+    key_len_bytes: usize,
+    rng: XorShiftRng,
+}
+
+impl RandomKeyGenerator {
+    pub fn new(key_bytes: usize) -> Self {
+        Self {
+            key_len_bytes: key_bytes,
+            rng: rand_xorshift::XorShiftRng::from_entropy(),
+        }
+    }
+}
+
+impl KeyGenerator for RandomKeyGenerator {
+    fn next_key(&mut self) -> Bytes {
+        let mut bytes = vec![0u8; self.key_len_bytes];
+        self.rng.fill_bytes(bytes.as_mut_slice());
+        Bytes::copy_from_slice(bytes.as_slice())
+    }
+}
+
+/// The database benchmarker.
+pub struct DbBench {
+    key_gen_supplier: Box<dyn Fn() -> Box<dyn KeyGenerator>>,
+    val_len: usize,
+    write_options: WriteOptions,
+    concurrency: u32,
+    num_rows: Option<u64>,
+    duration: Option<Duration>,
+    put_percentage: u32,
+    db: Arc<Db>,
+}
+
+impl DbBench {
+    // Ignore too many args
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        key_gen_supplier: Box<dyn Fn() -> Box<dyn KeyGenerator>>,
+        val_len: usize,
+        write_options: WriteOptions,
+        concurrency: u32,
+        num_rows: Option<u64>,
+        duration: Option<Duration>,
+        put_percentage: u32,
+        db: Arc<Db>,
+    ) -> Self {
+        Self {
+            key_gen_supplier,
+            val_len,
+            write_options,
+            concurrency,
+            num_rows,
+            duration,
+            put_percentage,
+            db,
+        }
+    }
+
+    /// Run the benchmarker.
+    ///
+    /// This method spins up `concurrency` tasks, each of which runs a loop,
+    /// and then waits for all the tasks to complete. It also spawns a task
+    /// to dump stats to the console.
+    pub async fn run(&self) {
+        let stats_recorder = Arc::new(StatsRecorder::new());
+        let mut write_tasks = Vec::new();
+        for _ in 0..self.concurrency {
+            let mut write_task = Task::new(
```

Comment:
nit: rename to just `task` since its doing both reads and writes

Reply:
bfaa61e